content/lectures/lec03.md

```diff
@@ -7,4 +7,192 @@ slides: ""
 draft: false
 ---
 
-Coming soon!
+# Lecture Three: Docker Compose
+
+## Limitations of Docker as a production run system
+
+We've seen how useful Docker is as a tool and the reasons that we'd want to use it in production. But it doesn't solve all of our problems. Take this `docker run` invocation, for example:
+
+```
+$ docker run -it \
+  --name=letsencrypt \
+  --cap-add=NET_ADMIN \
+  -e PUID=1000 \
+  -e PGID=1000 -e TZ=Europe/London \
+  -e URL=walthome.duckdns.org \
+  -e SUBDOMAINS=wildcard \
+  -e VALIDATION=duckdns \
+  -e DUCKDNSTOKEN=youwillnevergetthis \
+  -e EMAIL=pwpon500@gmail.com \
+  -v ./test:/config \
+  --restart unless-stopped \
+  linuxserver/letsencrypt
+```
+
+Maybe this is easy for the computer to read, but for humans there are numerous drawbacks to writing long console commands like this. For one, they're inaccessible and confusing to people who have less experience with the Docker CLI. Second, how are we going to keep track of this? Perhaps we could throw it in a shell script and commit it to our repository, but shell scripts can be cumbersome to edit and do not always interface well with other programming languages.
+
+## Run System Wishlist
+
+So, if we're unhappy with the Docker CLI as our interface for running containers, let's think about what we want out of a run system:
+
+- Reproducible (should be easy to share)
+- Easy to understand what's going on (human readable)
+- Confidence in making changes to config
+- Resource lifecycle management (we can set parameters that manage our containers)
```

Comment:
lifecycle is concerned with the action of creating and cleaning up containers after use, not managing which config is present for a container. without compose, you're forced to manually do all lifecycle management.